File: src/flashquant/core/packing.py
"""Nibble pack/unpack utilities for 4-bit index storage.

Packs pairs of 4-bit indices into single uint8 bytes to achieve 2x
storage reduction over one-byte-per-index representations. The packing
convention is: high nibble = even index, low nibble = odd index.

Tries the C++ extension ``_C`` first, falling back to pure PyTorch.

Examples:
    Pack and unpack round-trip::

        >>> indices = torch.randint(0, 16, (4, 128), dtype=torch.uint8)
        >>> packed = nibble_pack(indices)
        >>> packed.shape
        torch.Size([4, 64])
        >>> unpacked = nibble_unpack(packed)
        >>> (unpacked == indices.long()).all()
        True
"""

from __future__ import annotations

import logging

import torch

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Try C++ extension
# ---------------------------------------------------------------------------

_USE_C_PACKING = False
try:
    from flashquant import _C  # type: ignore[attr-defined]

    if hasattr(_C, "nibble_pack") and hasattr(_C, "nibble_unpack"):
        _USE_C_PACKING = True
        logger.debug("Using _C nibble pack/unpack backend")
except ImportError:
    pass

if not _USE_C_PACKING:
    logger.debug(
        "C++ packing extension not available; using pure PyTorch fallback"
    )


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def nibble_pack(indices: torch.Tensor) -> torch.Tensor:
    """Pack pairs of 4-bit indices into single uint8 bytes.

    Args:
        indices: uint8 tensor with values in [0, 15], last dimension
            must be even. Shape ``(..., D)``.

    Returns:
        uint8 tensor of shape ``(..., D // 2)`` with two indices
        per byte (high nibble = even index, low nibble = odd index).

    Raises:
        ValueError: If the last dimension is odd.
    """
    if indices.shape[-1] % 2 != 0:
        raise ValueError(
            f"Last dimension must be even for nibble packing, "
            f"got {indices.shape[-1]}"
        )

    if _USE_C_PACKING:
        return _C.nibble_pack(indices)  # type: ignore[attr-defined]

    even = indices[..., 0::2]
    odd = indices[..., 1::2]
    return ((even.to(torch.uint8) << 4) | odd.to(torch.uint8)).to(torch.uint8)


def nibble_unpack(packed: torch.Tensor) -> torch.Tensor:
    """Unpack nibble-packed uint8 bytes into pairs of 4-bit indices.

    Args:
        packed: uint8 tensor of shape ``(..., D // 2)`` with two
            indices per byte.

    Returns:
        Long tensor of shape ``(..., D)`` with individual indices
        suitable for centroid lookup.
    """
    if _USE_C_PACKING:
        return _C.nibble_unpack(packed)  # type: ignore[attr-defined]

    high = (packed >> 4).long()
    low = (packed & 0x0F).long()
    return torch.stack([high, low], dim=-1).flatten(-2)

File: csrc/core/packing.cpp
// Copyright 2026 FlashQuant Authors. Apache-2.0 License.
#include "packing.h"

#include <stdexcept>

namespace flashquant {

void validate_indices(const torch::Tensor& indices, int bits) {
    int max_val = (1 << bits) - 1;
    auto min_v = indices.min().item<int64_t>();
    auto max_v = indices.max().item<int64_t>();
    if (min_v < 0 || max_v > max_val) {
        throw std::invalid_argument(
            "Indices out of range for " + std::to_string(bits) +
            "-bit packing. Expected [0, " + std::to_string(max_val) +
            "], got [" + std::to_string(min_v) + ", " +
            std::to_string(max_v) + "]");
    }
}

torch::Tensor nibble_pack(const torch::Tensor& indices) {
    TORCH_CHECK(indices.size(-1) % 2 == 0,
                "Last dimension must be even for nibble packing, got ",
                indices.size(-1));

    validate_indices(indices, 4);

    auto idx = indices.to(torch::kUInt8);
    auto even = idx.index({torch::indexing::Ellipsis,
                           torch::indexing::Slice(0, torch::indexing::None, 2)});
    auto odd = idx.index({torch::indexing::Ellipsis,
                          torch::indexing::Slice(1, torch::indexing::None, 2)});

    return (even << 4) | odd;
}

torch::Tensor nibble_unpack(const torch::Tensor& packed) {
    auto high = (packed >> 4).to(torch::kInt32) & 0x0F;
    auto low = packed.to(torch::kInt32) & 0x0F;

    // Interleave: [h0, l0, h1, l1, ...]
    auto stacked = torch::stack({high, low}, -1);
    auto shape = packed.sizes().vec();
    shape.back() *= 2;
    return stacked.reshape(shape);
}

torch::Tensor pack_2bit(const torch::Tensor& indices) {
    TORCH_CHECK(indices.size(-1) % 4 == 0,
                "Last dimension must be divisible by 4 for 2-bit packing, got ",
                indices.size(-1));

    validate_indices(indices, 2);

    auto idx = indices.to(torch::kUInt8);
    auto v0 = idx.index({torch::indexing::Ellipsis,
                         torch::indexing::Slice(0, torch::indexing::None, 4)});
    auto v1 = idx.index({torch::indexing::Ellipsis,
                         torch::indexing::Slice(1, torch::indexing::None, 4)});
    auto v2 = idx.index({torch::indexing::Ellipsis,
                         torch::indexing::Slice(2, torch::indexing::None, 4)});
    auto v3 = idx.index({torch::indexing::Ellipsis,
                         torch::indexing::Slice(3, torch::indexing::None, 4)});

    return (v0 << 6) | (v1 << 4) | (v2 << 2) | v3;
}

torch::Tensor unpack_2bit(const torch::Tensor& packed, int64_t dim) {
    auto p = packed.to(torch::kInt32);
    auto v0 = (p >> 6) & 0x03;
    auto v1 = (p >> 4) & 0x03;
    auto v2 = (p >> 2) & 0x03;
    auto v3 = p & 0x03;

    auto stacked = torch::stack({v0, v1, v2, v3}, -1);
    auto shape = packed.sizes().vec();
    shape.back() = dim;
    return stacked.reshape(shape);
}

}  // namespace flashquant
